packages/backend-common/src/tokens/types.ts

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface TokenManager {
+  getServerToken: () => Promise<{ token: string }>;
+  validateServerToken: (token: string) => Promise<boolean>;
```

Comment:
Would like to avoid a boolean return here for two different reasons. One being evolvability and having this method return more things in the future, and the other being that it's a bit too big risk that users of this assume that it's going to throw and do things like this:

```ts
await this.tokenManager.validateServerToken(token)
return secrets
```

I think perhaps returning `void` for now and throwing an `AuthenticationError` will set us up somewhat nicely for future evolution?

Reply:
The method being async seems relevant to this discussion too. Seems like if we switch the return type to `Promise<void>` we end up with a similar anti-pattern where someone forgets to await the result. What do you think about making the method synchronous too to avoid that?

Reply:
Done (including proposed switch to sync validation) in 8a001f8202.

Reply:
Wanna push back on the switch to sync tbh, even though I totally agree that sync in isolation is better from a security point fo view. It's a somewhat important decision as it's tricky to switch between sync and async in the future and it limits the ways you are able to validate tokens, for example ruling out tokens stored in a DB.

One of the more important things to consider is that I think the server token management is likely to end up quite intertwined with the user token management, with some codepaths not really distinguishing between the two. For the user tokens it's a lot trickier to make that validation synchronous, as we're fetching signing certificates on demand there, or we may also want to reach out to check for revocations or store tokens in a DB instead. With the user token validation being async, I feel it makes sense to keep this one async too just to keep things similar.

Reply:
Totally agree that retaining the flexibility of async is desirable here. I have some reservations about switching to `async Promise<void>` though - it seems likely to lead to more incorrect usage than the original `async Promise<boolean>` API. I wonder if we could address the original concerns by instead renaming the `validateToken` method to something like `isTokenValid`?

Reply:
Yep, I think a return value is all good, just didn't wanna use boolean as it locks down the API completely and doesn't allow us to add any more returned information in the future.

What I'm thinking is that we eventually want something like this

```ts
const identity = await tokenManager.validateToken(token)
```

Do you think that makes sense and does it affect your thoughts on naming?  😁

Reply:
Ahh, that totally makes sense! I hadn't quite caught the fact that you were intending the `void` to be only a temporary situation. Your code snippet is kinda interesting - I was actually thinking separately that it would be awesome to be able to determine which plugin generated the token, so perhaps we can kill two birds with one stone here, create a BackstageBackendIdentity (or something) type which bakes in a pluginId, and return it from validateToken. What do you think?

Reply:
Yep exactly!

Only adding a plugin ID in isolation is a bit tricky as it would be trivial for a plugin to generate its own token with an arbitrary ID, but that's perhaps a challenge we can leave for later x)

Reply:
100% - had the same thought, this is a "self-reported" plugin ID. Pushed 4e6b2319a2 to the branch for this as a starting point, but looking forward to collaborating on this a bit.

Reply:
See [this comment](https://github.com/backstage/backstage/pull/7943#discussion_r757374312).